### PR TITLE
fix: querying on date partitions (fixes #1445)

### LINKED ISSUE #1445: Unable to query by partition column
# Environment
Max
**Delta-rs version**: latest(`deltalake = { git = "https://github.com/delta-io/delta-rs", branch = "main", features = ["s3", "datafusion"] }`)

**Binding**: 

**Environment**: OsX
- **Cloud provider**: AWS
- **OS**: Mac
- **Other**:

***
# Bug

**What happened**:

I create a table with a "date" partition column,

```
SchemaField::new(
            "date".to_string(),
            SchemaDataType::primitive("date".to_string()),
            false,
            HashMap::new(),
        );
```

```
DeltaOps(table)
                .create()
                .with_columns(delta_schema)
                .with_partition_columns(vec!["date"])
                .await?
```

If I run any query that either selects `date` or filters by `date`, I get:

```
Error: External error: External error: Arrow error: External error: Arrow error: Invalid argument error: column types must match schema types, expected Dictionary(UInt16, Date32) but found Dictionary(UInt16, Null) at column index 1

Caused by:
    0: External error: Arrow error: External error: Arrow error: Invalid argument error: column types must match schema types, expected Dictionary(UInt16, Date32) but found Dictionary(UInt16, Null) at column index 1
    1: Arrow error: External error: Arrow error: Invalid argument error: column types must match schema types, expected Dictionary(UInt16, Date32) but found Dictionary(UInt16, Null) at column index 1
    2: External error: Arrow error: Invalid argument error: column types must match schema types, expected Dictionary(UInt16, Date32) but found Dictionary(UInt16, Null) at column index 1
    3: Arrow error: Invalid argument error: column types must match schema types, expected Dictionary(UInt16, Date32) but found Dictionary(UInt16, Null) at column index 1
    4: Invalid argument error: column types must match schema types, expected Dictionary(UInt16, Date32) but found Dictionary(UInt16, Null) at column index 1
  ```

Removing the partition from the table and querying no longer results in an error, but querying on the date field is not working as expected (any query on the field results in no data). 

**What you expected to happen**:
I expect the queries to run and filter by date

**How to reproduce it**:
See above

### PR DESCRIPTION
**Pull request by ChewingGlass**
# Description
Fix querying on date partitions

# Related Issue(s)

- fixes #1445

### CODE REVIEW

File: rust/src/operations/transaction/state.rs [hunk 1]
```diff
@@ -52,6 +52,7 @@ impl DeltaTableState {
                                 // Dictionary encoding boolean types does not yield benefits
                                 // https://github.com/apache/arrow-datafusion/pull/5545#issuecomment-1526917997
                                 DataType::Boolean => field.data_type().clone(),
+                                DataType::Date32 => field.data_type().clone(),
                                 _ => wrap_partition_type_in_dict(field.data_type().clone()),
```
Comment (wjones127) [inline rust/src/operations/transaction/state.rs:56]: Let's fix this for more than dates:

```suggestion
                                // Only dictionary-encode types that may be large
                                // // https://github.com/apache/arrow-datafusion/pull/5545
                                DataType::Utf8 | DataType::LargeUtf8 | DataType::Binary | DataType::LargeBinary => wrap_partition_type_in_dict(field.data_type().clone()),
                                _ => field.data_type().clone()
```

File: rust/tests/datafusion_test.rs [hunk 2]
```diff
@@ -844,7 +844,7 @@ async fn test_issue_1292_datafusion_sql_projection() -> Result<()> {
 #[tokio::test]
 async fn test_issue_1291_datafusion_sql_partitioned_data() -> Result<()> {
     let ctx = SessionContext::new();
-    let table = deltalake::open_table("./tests/data/http_requests")
+    let table = deltalake::open_table("./tests/data/issue_1291")
```
[the remark this answers is not in the document]
Comment (wjones127) [reply rust/tests/datafusion_test.rs:847]: @ChewingGlass you willing to update this part soon?

[the remark this answers is not in the document]
Comment (ChewingGlass) [reply rust/tests/datafusion_test.rs:850]: Fixed it

[the remark this answers is not in the document]
Comment (wjones127) [reply rust/tests/datafusion_test.rs:850]: There seem to be some tests that need to be updated so they no longer expect dictionary encoded columns.
Also need to add the setup where you create a temporary directory and write the partitioned table :)

**Comment by watfordkcf** [comment]
What can I do to help move this along? We're currently pinning on a copy of this PR plus some merges and would very much like to move to a more official version.

**Comment by roeap** [comment]
@watfordkcf - seems like there are some tests that need fixing. Haven't had a chance to properly review yet, but it seems they are partially related to a table that was previously commited to the (within this PR) and later removed not being created within the tests.

Other failures are related to column encodings. On first glance related to how we encode the partitioned columns.

**Comment by wjones127** [comment]
Closing since we merged https://github.com/delta-io/delta-rs/pull/1594

**Pull request closed without merging**